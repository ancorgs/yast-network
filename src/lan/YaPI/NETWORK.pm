package YaPI::NETWORK;

use strict;
use YaST::YCP qw(:LOGGING);
use YaPI;
use Data::Dumper;
use Switch;

# ------------------- imported modules
YaST::YCP::Import ("LanItems");
YaST::YCP::Import ("Hostname");
YaST::YCP::Import ("DNS");
YaST::YCP::Import ("Routing");
# -------------------------------------

our $VERSION            = '1.0.0';
our @CAPABILITIES       = ('SLES11');
our %TYPEINFO;

# TODO: parameter map<string, boolean> what_I_Need
BEGIN{$TYPEINFO{Read} = ["function",
    [ "map", "string", "any"]];
}
sub Read {
  my $self	= shift;

# Hostname->Read();
 DNS->Read();
 Routing->Read();
 LanItems->Read();

 my %interfaces = ();
 foreach my $devnum (keys %{LanItems->Items}){
  LanItems->current($devnum);
  if (LanItems->IsItemConfigured()){
    my %configuration = ();
    LanItems->SetItem();
    if (LanItems->isCurrentDHCP()){
	%configuration = ( 'bootproto' => LanItems->bootproto );
    } elsif (LanItems->bootproto eq "static"){
	  %configuration = ( 'bootproto' => 'static' );
	  $configuration{'ipaddr'} = LanItems->ipaddr . "/" . LanItems->prefix;
	}
    $interfaces{LanItems->interfacename}=\%configuration;
  } elsif (LanItems->getCurrentItem()->{'hwinfo'}->{'type'} eq "eth") {
          my $device = LanItems->getCurrentItem()->{"hwinfo"}->{"dev_name"};
	  $interfaces{$device}= {};
	}
 }

  #FIXME: validate for nil values (dns espacially)
  my %ret	= ('interfaces'=>\%interfaces,
		   'routes'=>{'default'=>{'via'=>Routing->GetGateway()}}, 
                   'dns'=>{'nameservers'=>\@{DNS->nameservers}, 'searches'=>\@{DNS->searchlist}}, 
                   'hostname'=>{'name'=>Hostname->CurrentHostname, 'domain'=>Hostname->CurrentDomain}
		);
  return \%ret;
}

sub writeRoute {
    my $args  = shift;
    my %ret = ('exit'=>0, 'error'=>'');

    my $gw="";
    my $dest="";
    my @route = ();
    if (defined ($args->{'route'}->{'default'}->{'via'})){
      $gw = $args->{'route'}->{'default'}->{'via'};
      if ($gw ne ""){
        YaST::YCP::Import ("IP");
	unless (IP->Check4($gw)) {
	  $ret{'exit'} = -1;
	  $ret{'error'} = IP->Valid4();
	  return \%ret;	
	};
        $dest = "default";
        @route = ( {"destination" => $dest,
                        "gateway" => $gw,
                        "netmask" => "-",
                        "device" => "-"
                     });
      }
    }
    Routing->Read();
    y2milestone("YaPI->Write before change Routes:", Dumper(Routing->Routes));
    Routing->Routes( \@route );
    y2milestone("YaPI->Write after change Routes:", Dumper(Routing->Routes));
    Routing->Write();
  return \%ret;	
}

sub writeHostname {
   my $args  = shift;
   my $ret = {'exit'=>0, 'error'=>''};
   y2milestone("hostname", Dumper(\$args->{'hostname'}));
   DNS->Read();
   DNS->hostname($args->{'hostname'}->{'name'});
   DNS->domain($args->{'hostname'}->{'domain'});
   DNS->modified(1);
   DNS->Write();
  return $ret;
}

sub writeDNS {
   my $args  = shift;
   my $ret = {'exit'=>0, 'error'=>''};
   y2milestone("dns", Dumper(\$args->{'dns'}));
   DNS->Read();
   DNS->nameservers($args->{'dns'}->{'nameservers'});
   DNS->searchlist($args->{'dns'}->{'searches'});
   DNS->modified(1);
   DNS->Write();
  return $ret;
}

sub writeInterfaces {
   my $args  = shift;
   my $ret = {'exit'=>0, 'error'=>''};
   y2milestone("interface", Dumper(\$args->{'interface'}));
   foreach my $dev (keys %{$args->{'interface'}}){
           YaST::YCP::Import ("NetworkInterfaces");
           NetworkInterfaces->Read();
           NetworkInterfaces->Add() unless NetworkInterfaces->Edit($dev);
           NetworkInterfaces->Name($dev);
	   my $ip = $args->{'interface'}->{$dev}->{'ipaddr'};
           my $prefix="32";
           YaST::YCP::Import ("Netmask");
           my @ip_row = split(/\//, $ip);
           $prefix = $ip_row[$#ip_row];
           if (Netmask->Check4($prefix) && $prefix =~ /\./){
                y2milestone("Valid netmask: ", $prefix, " will change to prefixlen");
                $prefix = Netmask->ToBits($prefix);
           }
		$ip = $ip_row[0]."/".$prefix;
           my %config=("STARTMODE" => "auto",
                        "BOOTPROTO" => $args->{'interface'}->{$dev}->{'bootproto'},
                        "IPADDR" => $ip
                        );
           NetworkInterfaces->Current(\%config);
           NetworkInterfaces->Commit();
           NetworkInterfaces->Write("");
           YaST::YCP::Import ("Service");
	   Service->Restart("network");
   }
  return $ret;
}



BEGIN{$TYPEINFO{Write} = ["function",
    ["map","string","any"],["map","string","any"]];
}
sub Write {
  my $self  = shift;
  my $args  = shift;
  y2milestone("YaPI->Write with settings:", Dumper(\$args));

  # SAVE DEFAULT ROUTE
  if (exists($args->{'route'})){
    my $route_ret = writeRoute($args);
    return $route_ret if ($route_ret->{'exit'} != 0);
  }
  # SAVE HOSTNAME
  if (exists($args->{'hostname'})){
    my $hn_ret = writeHostname($args);
    return $hn_ret if ($hn_ret->{'exit'} != 0);
  }
  # SAVE DNS Settings
  if (exists($args->{'dns'})){
    my $dns_ret = writeDNS($args);
    return $dns_ret if ($dns_ret->{'exit'} != 0);
  }
  # SAVE interfaces Settings
  if (exists($args->{'interface'})){
    my $ifc_ret = writeInterfaces($args);
    return $ifc_ret if ($ifc_ret->{'exit'} != 0);
  }

 # return value for exit is type integer, but it'll be converted into string (in yast-perl-bindings)
 # that means in rest-api it'll be {'exit'=>'0', 'error'=>''}
 return {'exit'=>0, 'error'=>''};
}

1;
