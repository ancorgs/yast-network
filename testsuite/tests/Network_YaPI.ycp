{
include "testsuite.ycp";
import "YaPI::NETWORK";

map READ = $[
    "target" : $[
	"size":27,
	"string" : "laptop.suse.cz",
	"bash_output" : "laptop.suse.cz"
    ],
    "sysconfig": $[
        "network" : $[
	  "config" : $[
	    "NETCONFIG_DNS_STATIC_SERVERS" : "208.67.222.222 208.67.220.220",
	    "NETCONFIG_DNS_STATIC_SEARCHLIST" : "suse.cz suse.de"
	  ],
	  "dhcp" : $[
	    "DHCLIENT_SET_HOSTNAME" : "no",
	    "WRITE_HOSTNAME_TO_HOSTS" : "no"
	  ]
        ],
    ],
    "network" : $[
	  "section" : $[ "eth0":$[], "eth1":$[] ],
	  "value" : $[
		"eth0" : $[
		    "BOOTPROTO" : "dhcp4"
		  ],
	        "eth1" : $[
		    "BOOTPROTO" : "static",
		    "IPADDR"	: "1.2.3.4/24"
		  ]],
    ],
    "routes" : [ $["destination":"default", "gateway":"10.20.30.40"] ]
];

map EXEC = $[
    "target" : $[
	"bash_output" : $[
		"exit" : 0,
		"stdout" : "laptop.suse.cz"
		]
      ]
    ];

TEST(``(YaPI::NETWORK::Read()), [READ, $[], EXEC], nil);

TEST(``(YaPI::NETWORK::Write($[])), [$[], $[], $[]], nil);

// test correct default route
TEST(``(YaPI::NETWORK::Write($["route":$["default":$["via":"10.20.30.40"]]])), [READ], nil);

// test incorrect default route (invalid gw IP)
TEST(``(YaPI::NETWORK::Write($["route":$["default":$["via":"10.20.30"]]])), [READ], nil);
}
