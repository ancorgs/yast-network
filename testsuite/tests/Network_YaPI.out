Read	.target.tmpdir nil
Log	Failed to set temporary directory: nil
Read	.sysconfig.network.dhcp.DHCLIENT_SET_HOSTNAME "yes"
Read	.sysconfig.network.dhcp.WRITE_HOSTNAME_TO_HOSTS "no"
Read	.sysconfig.network.config.NETCONFIG_DNS_POLICY nil
Read	.sysconfig.network.config.NETCONFIG_DNS_STATIC_SERVERS "208.67.222.222 208.67.220.220"
Read	.sysconfig.network.config.NETCONFIG_DNS_STATIC_SEARCHLIST "suse.cz suse.de"
Read	.target.size "/etc/HOSTNAME" 27
Read	.target.string "/etc/HOSTNAME" "laptop.suse.cz"
Read	.target.size "/etc/sysconfig/network/routes" 27
Read	.routes [$["destination":"default", "gateway":"10.20.30.40"]]
Execute	.target.bash "rpm -q --whatprovides SuSEfirewall2" 0
Read	.init.scripts.exists "SuSEfirewall2_init" false
Read	.sysconfig.sysctl.IP_FORWARD nil
Dir	.network.section: ["eth0", "eth1", "eth2", "eth3", "eth4"]
Dir	.network.value."eth0": ["BOOTPROTO"]
Read	.network.value."eth0".BOOTPROTO "dhcp4"
Dir	.network.value."eth1": ["BOOTPROTO", "IPADDR"]
Read	.network.value."eth1".BOOTPROTO "static"
Read	.network.value."eth1".IPADDR "1.2.3.4/24"
Dir	.network.value."eth2": ["BOOTPROTO", "IPADDR", "PREFIXLEN"]
Read	.network.value."eth2".BOOTPROTO "static"
Read	.network.value."eth2".IPADDR "1.2.3.5/24"
Read	.network.value."eth2".PREFIXLEN ""
Dir	.network.value."eth3": ["BOOTPROTO", "IPADDR", "PREFIXLEN"]
Read	.network.value."eth3".BOOTPROTO "static"
Read	.network.value."eth3".IPADDR "1.2.3.6"
Read	.network.value."eth3".PREFIXLEN "24"
Dir	.network.value."eth4": ["BOOTPROTO", "IPADDR", "NETMASK"]
Read	.network.value."eth4".BOOTPROTO "static"
Read	.network.value."eth4".IPADDR "1.2.3.7"
Read	.network.value."eth4".NETMASK "255.255.255.0"
Read	.target.size "/etc/install.inf" 27
Dir	.etc.install_inf: []
Read	.target.string "/proc/cmdline" "laptop.suse.cz"
Read	.probe.netcard nil
Log	hardware detection failure
Read	.etc.install_inf.BrokenModules nil
Read	.udev_persistent.net nil
Dir	.modules.options: []
Read	.udev_persistent.drivers nil
Read	.target.lstat "/etc/sysconfig/network/ifcfg-eth-pcmcia" nil
Read	.target.lstat "/etc/sysconfig/network/ifcfg-eth-usb" nil
Read	.target.lstat "/etc/sysconfig/network/ifcfg-tr-pcmcia" nil
Read	.target.lstat "/etc/sysconfig/network/ifcfg-tr-usb" nil
Execute	.target.bash_output "hostname --fqdn" $["exit":0, "stdout":"laptop.suse.cz"]
Execute	.target.bash_output "hostname --fqdn" $["exit":0, "stdout":"laptop.suse.cz"]
Return	$["dns":$["nameservers":["208.67.222.222", "208.67.220.220"], "searches":["suse.cz", "suse.de"]], "hostname":$["dhcp_hostname":"1", "domain":"suse.cz", "name":"laptop"], "interfaces":$["eth0":$["bootproto":"dhcp4"], "eth1":$["bootproto":"static", "ipaddr":"1.2.3.4/24"], "eth2":$["bootproto":"static", "ipaddr":"1.2.3.5/24"], "eth3":$["bootproto":"static", "ipaddr":"1.2.3.6/24"], "eth4":$["bootproto":"static", "ipaddr":"1.2.3.7/24"]], "routes":$["default":$["via":"10.20.30.40"]]]
Return	$["error":"", "exit":"0"]
Read	.target.size "/etc/sysconfig/network/routes" 27
Read	.routes [$["destination":"default", "gateway":"10.20.30.40"]]
Read	.init.scripts.exists "SuSEfirewall2_init" false
Read	.sysconfig.sysctl.IP_FORWARD nil
Read	.init.scripts.exists "SuSEfirewall2_init" false
Write	.sysconfig.sysctl.IP_FORWARD "no" true
Write	.sysconfig.sysctl.IPV6_FORWARD "no" true
Write	.sysconfig.sysctl nil true
Execute	.target.bash "echo 0 > /proc/sys/net/ipv4/ip_forward" 0
Execute	.target.bash "echo 0 > /proc/sys/net/ipv6/conf/all/forwarding" 0
Read	.target.size "/etc/sysconfig/network/routes" 27
Execute	.target.bash "/bin/cp /etc/sysconfig/network/routes /etc/sysconfig/network/routes.YaST2save" 0
Write	.routes [$["destination":"default", "device":"-", "gateway":"10.20.30.40", "netmask":"-"]] true
Return	$["error":"", "exit":"0"]
Return	$["error":"A valid IP address consists of four integers\nin the range 0-255 separated by dots.", "exit":"-1"]
Write	.network.value."eth3".BOOTPROTO "static" true
Write	.network.value."eth3".IPADDR "1.2.3.7/24" true
Write	.network.value."eth3".STARTMODE "auto" true
Write	.network nil true
Read	.init.scripts.exists "network" false
Return	$["error":"", "exit":"0"]
Read	.target.size "/etc/sysconfig/network/routes" 27
Read	.routes [$["destination":"default", "gateway":"10.20.30.40"]]
Read	.init.scripts.exists "SuSEfirewall2_init" false
Read	.sysconfig.sysctl.IP_FORWARD nil
Read	.probe.netcard nil
Log	hardware detection failure
Read	.etc.install_inf.BrokenModules nil
Read	.udev_persistent.net nil
Dir	.modules.options: []
Read	.udev_persistent.drivers nil
Read	.target.lstat "/etc/sysconfig/network/ifcfg-eth-pcmcia" nil
Read	.target.lstat "/etc/sysconfig/network/ifcfg-eth-usb" nil
Read	.target.lstat "/etc/sysconfig/network/ifcfg-tr-pcmcia" nil
Read	.target.lstat "/etc/sysconfig/network/ifcfg-tr-usb" nil
Execute	.target.bash_output "hostname --fqdn" $["exit":0, "stdout":"laptop.suse.cz"]
Execute	.target.bash_output "hostname --fqdn" $["exit":0, "stdout":"laptop.suse.cz"]
Return	$["dns":$["nameservers":["208.67.222.222", "208.67.220.220"], "searches":["suse.cz", "suse.de"]], "hostname":$["dhcp_hostname":"1", "domain":"suse.cz", "name":"laptop"], "interfaces":$["eth0":$["bootproto":"dhcp4"], "eth1":$["bootproto":"static", "ipaddr":"1.2.3.4/24"], "eth2":$["bootproto":"static", "ipaddr":"1.2.3.5/24"], "eth3":$["bootproto":"static", "ipaddr":"1.2.3.7/24/"], "eth4":$["bootproto":"static", "ipaddr":"1.2.3.7/24"]], "routes":$["default":$["via":"10.20.30.40"]]]
